### ISSUE TITLE
[BUG] Limited API: generate_type_import_call() uses opaque type PyTypeObject

### ISSUE THREAD

**Issue by aws-taylor** [cython/cython#5634]
### Describe the bug

The `generate_type_import_call` methods includes calls to `sizeof(PyTypeObject)` that results in compilation failures if the limited API is enabled - https://github.com/cython/cython/blob/d455d51bf31379f47c074e40517e24857d4d9cc0/Cython/Compiler/ModuleNode.py#L3859

Specifically, you get:
```
error: invalid application of ‘sizeof’ to incomplete type ‘PyTypeObject’ {aka ‘_typeobject’}
 3076 |   sizeof(PyTypeObject), __PYX_GET_STRUCT_ALIGNMENT_3_0_0(PyTypeObject),
      |   ^~~~~~~~~~~~~~~~~~~~
```

And:

```
error: invalid application of ‘__alignof__’ to incomplete type ‘PyTypeObject’ {aka ‘_typeobject’}
 3076 |   sizeof(PyTypeObject), __PYX_GET_STRUCT_ALIGNMENT_3_0_0(PyTypeObject),
 ```

The generated code is:
`
__pyx_ptype_7cpython_4type_type = __Pyx_ImportType_3_0_0(__pyx_t_1, __Pyx_BUILTIN_MODULE_NAME, "type", sizeof(PyTypeObject), __PYX_GET_STRUCT_ALIGNMENT_3_0_0(PyTypeObject), __Pyx_ImportType_CheckSize_Warn_3_0_0);`
 
 
  I'm confused by this code in general because `PyTypeObject` and `PyType_Type` are directly available in the limited API, so it's not clear to me why it's necessary to call `__Pyx_ImportType_3_0_0` in the first place.

**Comment by da-woods** [comment]
This feels like another thing caused by a stray cimport. I can get this code from

```
from cpython cimport type
```

I'm not completely confident that there isn't another way to get it, but we generate `__Pyx_ImportType_3_0_0` when cimporting types.

**Comment by aws-taylor** [comment]
Yup,  `from cpython.bytes cimport PyBytes_AsString` -> `from .object cimport PyObject` -> `cimport cpython.type`

I was able to work around the issue by cherry-picking the specific symbols I needed:

```
cdef extern from "Python.h":
  char* PyBytes_AsString(object string) except NULL
```

I'm still confused by the implementation though - why is `__Pyx_ImportType_3_0_0` necessary?

**Comment by da-woods** [comment]
> I'm still confused by the implementation though - why is `__Pyx_ImportType_3_0_0` necessary?

type.pxd includes `type` as an extern cdef class:

https://github.com/cython/cython/blob/master/Cython/Includes/cpython/type.pxd

It's largely just defined as opaque, but the idea is that you'd be able to inherit a cdef class from it:

```
cdef class SomeExtendedType(cpython.type.type):
    # ...
```

The check is that the size of `type` matches the size that Cython believes `type` to be. If it isn't then you end up with a binary compatibility mess. That's obviously more of an issue for when cimporting from 3rd party modules rather than builtin types, but the same logic applies to every cimported type.

*However*, it may all be pointless in this case. Cython already knows what `type` is without a `cimport` so you can already do

```
# Note that this will probably never work in the limited API, but it's just an illustration...)
cdef class T(type):
    pass
```

Therefore, I'm not really sure why we `type` defined as an extern class in this case since it doesn't give you access to anything new. We probably can't removeit from `cpython.type` easily without breaking existing code. But I don't think `cpython.object` needs to import it.

I think all this `cimport cpython` files are going to need some real thought about how they become limited-API compatible.

**Comment by aws-taylor** [comment]
Aha, that all makes sense. Thanks. 

It looks like the only usage of `type` in all of the .pxd files is:
`ctypedef object (*newfunc)(cpython.type.type, PyObject*, PyObject*)  # (type, args|NULL, kwargs|NULL)`

Can this be replaced with?
`ctypedef object (*newfunc)(type, PyObject*, PyObject*)  # (type, args|NULL, kwargs|NULL)` 
Or:
`ctypedef object (*newfunc)(PyTypeObject, PyObject*, PyObject*)  # (type, args|NULL, kwargs|NULL)`

**Comment by da-woods** [comment]
> Can this be replaced with?
`ctypedef object (*newfunc)(type, PyObject*, PyObject*)  # (type, args|NULL, kwargs|NULL)`

I think so. Would want to do a bit of testing to check it didn't break existing code though, but I think that should work
